In [1]:
import sys
import os

sys.path.append(os.path.abspath(".."))

In [ ]:
import os
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import DataLoader

from models.unetpp_3d import UNetPP3D
from utils.unets_helper_functions import (
    set_seed,
    train_one_epoch,
    validate_one_epoch,
    save_checkpoint,
    PatchDataset,
    evaluate_full_volume
)

In [3]:
set_seed(42)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
print("Using device:", device)


Using device: cuda


In [4]:
FOLD = 0

with open(f"../data/splits/fold_{FOLD}/train.txt") as f:
    train_cases = f.read().splitlines()

with open(f"../data/splits/fold_{FOLD}/val.txt") as f:
    val_cases = f.read().splitlines()

print("Train cases:", len(train_cases))
print("Val cases:", len(val_cases))


Train cases: 33
Val cases: 9


In [5]:
PATCH_SIZE = 80
PATCHES_PER_CASE = 6

train_dataset = PatchDataset(train_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = PATCHES_PER_CASE,
                        patch_size = PATCH_SIZE,
                        augment=False)

val_dataset = PatchDataset(val_cases,
                        "../data/processed/imagesTr",
                        "../data/processed/labelsTr",
                        patches_per_case = 1,
                        patch_size = PATCH_SIZE,
                        augment=False)


train_loader = DataLoader(
    train_dataset,
    batch_size=2,
    shuffle=True,
    num_workers=2,      
    pin_memory=True    
)

val_loader = DataLoader(
    val_dataset,
    batch_size=1,
    num_workers=2,    
    pin_memory=True
)

In [6]:
model = UNetPP3D(
    in_channels=1,
    out_channels=7,
    base_filters=16
).to(device)

print("UNet++ Initialized")

UNet++ Initialized


In [7]:
optimizer = optim.Adam(model.parameters(), lr=1e-4)
weights = torch.tensor([0.1, 1, 1, 1, 1, 1, 1]).to(device)
ce_loss = nn.CrossEntropyLoss(weight=weights)
scaler = torch.cuda.amp.GradScaler()


C:\Users\dhanu\AppData\Local\Temp\ipykernel_17192\2337973256.py:4: FutureWarning: `torch.cuda.amp.GradScaler(args...)` is deprecated. Please use `torch.amp.GradScaler('cuda', args...)` instead.
  scaler = torch.cuda.amp.GradScaler()


In [8]:
EPOCHS = 40

PROJECT_ROOT = os.path.abspath("..")
SAVE_DIR = os.path.join(PROJECT_ROOT, "experiments", "unetpp_fold0")

os.makedirs(SAVE_DIR, exist_ok=True)
best_val_loss = float("inf")

best_val_loss = float("inf")

for epoch in range(EPOCHS):

    train_loss = train_one_epoch(
        model,
        train_loader,
        optimizer,
        scaler,
        ce_loss,
        device
    )

    val_loss, val_dice = validate_one_epoch(
        model,
        val_loader,
        ce_loss,
        device
    )

    print(f"\nEpoch {epoch}")
    print(f"Train Loss: {train_loss:.4f}")
    print(f"Val Loss: {val_loss:.4f}")
    print(f"Per Class Dice: {val_dice}")

    # Save best model
    if val_loss < best_val_loss:
        best_val_loss = val_loss
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, "best_model.pth")
        )
        print("Saved Best Model")

    # Save checkpoint every 10 epochs
    if (epoch + 1) % 10 == 0:
        save_checkpoint(
            model,
            optimizer,
            epoch,
            best_val_loss,
            os.path.join(SAVE_DIR, f"checkpoint_epoch_{epoch+1}.pth")
        )
        print("Checkpoint Saved")

100%|██████████| 99/99 [06:15<00:00,  3.79s/it]



Epoch 0
Train Loss: 2.5323
Val Loss: 2.3817
Per Class Dice: [0.21326228 0.08110614 0.01755217 0.09770568 0.20773341 0.11496629]
Saved Best Model


100%|██████████| 99/99 [05:24<00:00,  3.28s/it]



Epoch 1
Train Loss: 2.3489
Val Loss: 2.2711
Per Class Dice: [0.28215537 0.06424229 0.02336434 0.05234946 0.12855055 0.10515272]
Saved Best Model


100%|██████████| 99/99 [05:39<00:00,  3.43s/it]



Epoch 2
Train Loss: 2.1999
Val Loss: 2.2469
Per Class Dice: [0.44640684 0.06536265 0.0319449  0.10369543 0.10288083 0.08120145]
Saved Best Model


100%|██████████| 99/99 [05:44<00:00,  3.48s/it]



Epoch 3
Train Loss: 2.0991
Val Loss: 1.9799
Per Class Dice: [0.44037985 0.20128361 0.11507709 0.24669588 0.16730257 0.29486122]
Saved Best Model


100%|██████████| 99/99 [05:42<00:00,  3.46s/it]



Epoch 4
Train Loss: 2.0294
Val Loss: 1.9557
Per Class Dice: [0.40355389 0.38694263 0.23666232 0.39572342 0.24417982 0.21310709]
Saved Best Model


100%|██████████| 99/99 [05:52<00:00,  3.56s/it]



Epoch 5
Train Loss: 1.9805
Val Loss: 1.9426
Per Class Dice: [0.49939655 0.37060295 0.19549352 0.29338232 0.42832127 0.23648598]
Saved Best Model


100%|██████████| 99/99 [06:27<00:00,  3.91s/it]



Epoch 6
Train Loss: 1.8854
Val Loss: 1.7531
Per Class Dice: [0.74900869 0.4021399  0.35474362 0.63491901 0.50679891 0.58543232]
Saved Best Model


100%|██████████| 99/99 [05:23<00:00,  3.27s/it]



Epoch 7
Train Loss: 1.8386
Val Loss: 1.8132
Per Class Dice: [0.56048748 0.38045831 0.3817584  0.42467532 0.40197636 0.33863954]


100%|██████████| 99/99 [06:03<00:00,  3.68s/it]



Epoch 8
Train Loss: 1.7603
Val Loss: 1.7200
Per Class Dice: [0.49991055 0.34115033 0.18415568 0.59409506 0.6963577  0.4676963 ]
Saved Best Model


100%|██████████| 99/99 [06:01<00:00,  3.65s/it]



Epoch 9
Train Loss: 1.7092
Val Loss: 1.6047
Per Class Dice: [0.81714947 0.63341787 0.45332574 0.70664734 0.79133637 0.71796925]
Saved Best Model
Checkpoint Saved


100%|██████████| 99/99 [04:50<00:00,  2.93s/it]



Epoch 10
Train Loss: 1.6621
Val Loss: 1.6911
Per Class Dice: [0.60806375 0.50522267 0.54338103 0.4392634  0.57761415 0.36726039]


100%|██████████| 99/99 [04:41<00:00,  2.84s/it]



Epoch 11
Train Loss: 1.6182
Val Loss: 1.6594
Per Class Dice: [0.61385257 0.54271569 0.56480822 0.59318284 0.79865872 0.6346409 ]


100%|██████████| 99/99 [05:15<00:00,  3.18s/it]



Epoch 12
Train Loss: 1.5694
Val Loss: 1.5373
Per Class Dice: [0.74563381 0.63393992 0.52477527 0.71345295 0.88149642 0.64786547]
Saved Best Model


100%|██████████| 99/99 [05:58<00:00,  3.62s/it]



Epoch 13
Train Loss: 1.5297
Val Loss: 1.5915
Per Class Dice: [0.77180944 0.7777781  0.66666669 0.74177452 0.88888889 0.76247407]


100%|██████████| 99/99 [05:53<00:00,  3.57s/it]



Epoch 14
Train Loss: 1.5009
Val Loss: 1.5022
Per Class Dice: [0.60746709 0.66420147 0.56906733 0.73476234 0.77910862 0.62821636]
Saved Best Model


100%|██████████| 99/99 [05:36<00:00,  3.40s/it]



Epoch 15
Train Loss: 1.4622
Val Loss: 1.3976
Per Class Dice: [0.82095685 0.80643675 0.72153125 0.83690215 0.79421314 0.7829739 ]
Saved Best Model


100%|██████████| 99/99 [05:23<00:00,  3.27s/it]



Epoch 16
Train Loss: 1.4170
Val Loss: 1.4078
Per Class Dice: [0.72077768 0.44938774 0.49412884 0.52262301 0.63028068 0.53690875]


100%|██████████| 99/99 [05:56<00:00,  3.60s/it]



Epoch 17
Train Loss: 1.3745
Val Loss: 1.3762
Per Class Dice: [0.61184295 0.64236385 0.7674918  0.60595743 0.72108967 0.59533728]
Saved Best Model


100%|██████████| 99/99 [05:24<00:00,  3.27s/it]



Epoch 18
Train Loss: 1.3519
Val Loss: 1.3324
Per Class Dice: [0.86961136 0.74097645 0.72757325 0.80787246 0.8539914  0.83261045]
Saved Best Model


100%|██████████| 99/99 [05:55<00:00,  3.59s/it]



Epoch 19
Train Loss: 1.3253
Val Loss: 1.2967
Per Class Dice: [0.77805527 0.70159308 0.82101991 0.74068544 0.69897313 0.71809311]
Saved Best Model
Checkpoint Saved


100%|██████████| 99/99 [05:34<00:00,  3.37s/it]



Epoch 20
Train Loss: 1.3034
Val Loss: 1.2753
Per Class Dice: [0.77629438 0.65916323 0.91324432 0.62020314 0.97174164 0.87100381]
Saved Best Model


100%|██████████| 99/99 [05:32<00:00,  3.36s/it]



Epoch 21
Train Loss: 1.2827
Val Loss: 1.2903
Per Class Dice: [0.71231607 0.86750074 0.94107868 0.73840204 0.56289806 0.71203211]


100%|██████████| 99/99 [05:33<00:00,  3.37s/it]



Epoch 22
Train Loss: 1.2476
Val Loss: 1.2773
Per Class Dice: [0.60987422 0.49936592 0.64181527 0.54538726 0.89203196 0.55221614]


100%|██████████| 99/99 [05:56<00:00,  3.60s/it]



Epoch 23
Train Loss: 1.2186
Val Loss: 1.2184
Per Class Dice: [0.57909677 0.61029899 0.60647156 0.57062897 0.64534511 0.64878246]
Saved Best Model


100%|██████████| 99/99 [06:07<00:00,  3.71s/it]



Epoch 24
Train Loss: 1.1908
Val Loss: 1.1124
Per Class Dice: [0.78264659 0.84061336 0.63654757 0.7180675  0.84032931 0.75157888]
Saved Best Model


100%|██████████| 99/99 [06:25<00:00,  3.90s/it]



Epoch 25
Train Loss: 1.1717
Val Loss: 1.1871
Per Class Dice: [0.51929531 0.67251662 0.67323835 0.59617834 0.95811925 0.92539769]


100%|██████████| 99/99 [06:39<00:00,  4.03s/it]



Epoch 26
Train Loss: 1.1547
Val Loss: 1.1484
Per Class Dice: [0.7575677  0.65932649 0.58147727 0.91884682 0.76349254 0.80894992]


100%|██████████| 99/99 [05:19<00:00,  3.22s/it]



Epoch 27
Train Loss: 1.1379
Val Loss: 1.1614
Per Class Dice: [0.83822478 0.90701474 0.94913613 0.84514504 0.9355628  0.80014725]


100%|██████████| 99/99 [04:56<00:00,  3.00s/it]



Epoch 28
Train Loss: 1.1306
Val Loss: 1.1172
Per Class Dice: [0.50209032 0.57552347 0.88637959 0.54946169 0.88130965 0.69779784]


100%|██████████| 99/99 [04:41<00:00,  2.84s/it]



Epoch 29
Train Loss: 1.1024
Val Loss: 1.1171
Per Class Dice: [0.80302945 0.79876355 0.79688568 0.44233184 0.75710801 0.78713376]
Checkpoint Saved


100%|██████████| 99/99 [04:25<00:00,  2.68s/it]



Epoch 30
Train Loss: 1.0947
Val Loss: 1.1097
Per Class Dice: [0.967367   0.94276335 0.83818296 0.8865934  0.88888889 0.82715468]
Saved Best Model


100%|██████████| 99/99 [04:50<00:00,  2.93s/it]



Epoch 31
Train Loss: 1.0593
Val Loss: 1.0603
Per Class Dice: [0.81403799 0.72143921 0.78722234 0.64096881 0.68027373 0.66502479]
Saved Best Model


100%|██████████| 99/99 [04:16<00:00,  2.59s/it]



Epoch 32
Train Loss: 1.0612
Val Loss: 1.0261
Per Class Dice: [0.70773025 0.70558213 0.7752501  0.72485215 0.79289178 0.69094082]
Saved Best Model


100%|██████████| 99/99 [04:24<00:00,  2.67s/it]



Epoch 33
Train Loss: 1.0510
Val Loss: 1.0264
Per Class Dice: [0.66443993 0.56239009 0.60952535 0.70506379 0.57494309 0.57338898]


100%|██████████| 99/99 [04:29<00:00,  2.72s/it]



Epoch 34
Train Loss: 1.0375
Val Loss: 1.0591
Per Class Dice: [0.66636038 0.70829001 0.96701802 0.72101998 0.84686884 0.73155491]


100%|██████████| 99/99 [04:43<00:00,  2.86s/it]



Epoch 35
Train Loss: 1.0059
Val Loss: 0.9830
Per Class Dice: [0.68360723 0.50822322 0.76378515 0.65650279 0.61310745 0.66580378]
Saved Best Model


100%|██████████| 99/99 [04:29<00:00,  2.72s/it]



Epoch 36
Train Loss: 1.0071
Val Loss: 0.9883
Per Class Dice: [0.70376233 0.89537846 0.73820886 0.74832204 0.77077939 0.72887146]


100%|██████████| 99/99 [04:51<00:00,  2.94s/it]



Epoch 37
Train Loss: 0.9649
Val Loss: 1.0179
Per Class Dice: [0.7249639  0.93703436 0.90044845 0.79180059 0.82530532 0.92849108]


100%|██████████| 99/99 [04:27<00:00,  2.71s/it]



Epoch 38
Train Loss: 0.9846
Val Loss: 0.9172
Per Class Dice: [0.815478   0.67810958 0.80788818 0.73203245 0.88476692 0.78494775]
Saved Best Model


100%|██████████| 99/99 [04:39<00:00,  2.83s/it]



Epoch 39
Train Loss: 0.9655
Val Loss: 1.0260
Per Class Dice: [0.73134724 0.94395071 0.90352526 0.53249118 0.9139178  0.93729089]
Checkpoint Saved


In [9]:
print("Training Complete")

Training Complete


In [5]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

model = UNetPP3D().to(device)

model_path = "../experiments/unetpp_fold0/best_model.pth"
checkpoint = torch.load(model_path, map_location=device)
model.load_state_dict(checkpoint["model_state_dict"])

model.eval()

print("Model loaded for testing.")


Model loaded for testing.


C:\Users\dhanu\AppData\Local\Temp\ipykernel_18528\2368241481.py:6: FutureWarning: You are using `torch.load` with `weights_only=False` (the current default value), which uses the default pickle module implicitly. It is possible to construct malicious pickle data which will execute arbitrary code during unpickling (See https://github.com/pytorch/pytorch/blob/main/SECURITY.md#untrusted-models for more details). In a future release, the default value for `weights_only` will be flipped to `True`. This limits the functions that could be executed during unpickling. Arbitrary objects will no longer be allowed to be loaded via this mode unless they are explicitly allowlisted by the user via `torch.serialization.add_safe_globals`. We recommend you start setting `weights_only=True` for any use case where you don't have full control of the loaded file. Please open an issue on GitHub for any issues related to this experimental feature.
  checkpoint = torch.load(model_path, map_location=device)


In [6]:
mean_dice, std_dice = evaluate_full_volume(
    model,
    val_cases,
    "../data/processed/imagesTr",
    "../data/processed/labelsTr",
    device=device
)

Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_38 Dice: [np.float64(0.7619944738001352), np.float64(0.7279124817811321), np.float64(0.5383188246919954), np.float64(0.6886213947394796), np.float64(0.7246009629891048), np.float64(0.3332003651443021)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_05 Dice: [np.float64(0.7224811184953327), np.float64(0.6280422951710929), np.float64(0.6565396502876678), np.float64(0.7778071335388074), np.float64(0.8239129170359666), np.float64(0.37735180965051573)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_22 Dice: [np.float64(0.6793798811797692), np.float64(0.7291597293100457), np.float64(0.5304404755162813), np.float64(0.718290617127553), np.float64(0.7452416693213799), np.float64(0.3461536124658232)]
Unique pred labels: [0 1 2 3 4 5 6]
Unique GT labels: [0 1 2 3 4 5 6]
case_14 Dice: [np.float64(0.7335535302521783), np.float64(0.7627708545173623), np.float64(0.5608818

## UNet++ 3D Patch-Based Validation Performance

The model was trained for 40 epochs using patch-based validation.  
Below are the final patch validation metrics obtained at the best validation loss epoch.

### Final Training Statistics (Epoch 38 - Best Model)

- Train Loss: 0.9846  
- Validation Loss: 0.9172  

### Patch-Based Per-Class Dice (Validation)

- Mandible: 0.8155  
- Brainstem: 0.6781  
- Spinal Cord: 0.8079  
- Parotid Left: 0.7320  
- Parotid Right: 0.8848  
- Oral Cavity: 0.7849  


## Observations

- Parotid glands and Mandible achieve high Dice scores (>0.80).
- Spinal Cord performance is strong under patch-based validation.

Overall, the UNet++ model demonstrates effective multi-organ segmentation performance under patch-based validation.


---
# Evaluating on Full Volume 
~ (patch = 80 , kernel = 60)

| Organ ID | Organ (Based on Your Map) | Mean Dice  | Std Dev |
| -------- | ------------------------- | ---------- | ------- |
| 1        | Bone Mandible             | **0.7615** | 0.0455  |
| 2        | Brainstem                 | **0.7197** | 0.0520  |
| 3        | Spinal Cord               | **0.5555** | 0.0460  |
| 4        | Parotid Left              | **0.7037** | 0.0480  |
| 5        | Parotid Right             | **0.7239** | 0.0825  |
| 6        | Oral Cavity               | **0.3755** | 0.0438  |


Full Volume eval Summary

**Strong Organs**
Mandible (0.76) → Very stable and strong
Brainstem (0.72) → Consistent
Parotids (~0.70–0.72) → Good bilateral symmetry

**Moderate**
Spinal Cord (0.55)
This is actually reasonable for 3D patch training
Thin elongated structures are hardest in sliding-window inference

**Weakest**
Oral Cavity (0.37)
Likely due to:
Irregular shape, High anatomical variation ,Lower contrast boundary